framework/src/mesh/MooseMesh.C

```diff
@@ -728,6 +736,30 @@ MooseMesh::nodeToElemMap()
   return _node_to_elem_map;
 }
 
+const std::unordered_map<dof_id_type, std::vector<const Elem *>> &
+MooseMesh::nodeToElemPtrMap()
+{
+  if (!_node_to_elem_ptr_map_built) // Guard the creation with a double checked lock
+  {
+    Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
+
+    if (!_node_to_elem_ptr_map_built)
+    {
+      TIME_SECTION(_node_to_elem_ptr_map_timer);
+      CONSOLE_TIMED_PRINT("Building node to element pointer map");
+
+      for (const auto & elem : getMesh().active_element_ptr_range())
```

Comment:
First, what prompted you to work on this ticket? I'm asking because I'm guessing you ran into an issue somewhere? You didn't really work on this just because you saw a pattern where the caller had to get the actual element right? Anyway, this is besides the point. We might want to consider cleaning this map up. With the way it stands now, we are building a global map on every processor when using ReplicatedMesh. Based on the usage that you updated, I'm not convinced that there are any cases where we need non-local information. We certainly don't need non-ghosted information anywhere. 

This method is quite old and we could probably loop over just the ghosted node range and build a partial map that's complete enough for all use cases. The reason I say this with certainty is this clearly isn't breaking anyone who is running with distributed mesh which would already be building only a partial map of information (i.e. the active elements are not complete in DistributedMesh mode). Maybe just try it?

Reply:
For context: in ray tracing, specifically spatial claiming, we end up needing a map of elements to all of their vertex neighbors. This is built by first building a node to element map and then using this info to make the element to vertex neighbor map. I compared what I did there to what's done in `MooseMesh` and was surprised to see that we built the node to elem map as such. I plan on adding said element to vertex neighbor map later as well.

Back to the main point - I think this is a good idea. More so because I'm already working on it. I'll give this ghosted range a try and see if it breaks any of the framework use cases that I changed here.

Reply:
We want to do _all_ of the elements (unless we have a great way to do all the non-local element pulled in by RelationshipManagers).

We need these maps to work for DistributedMesh with geometric search... where geometric search has ghosted over "random" elements (which wouldn't be in the "ghosted range").

Reply:
To fix the ReplicatedMesh problem: just don't use replicated mesh...

Reply:
How about both, then? 

`nodeToElemPtr()` and `ghostedNodeToElemPtr()` ?

Reply:
The consensus here was to have `nodeToElemPtr()` default to evaluable elems, and add a boolean for using geometric elems instead